File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit= 0, num=prices[0];
        int n = prices.size();

        for(int i=1; i<n; i++){

          
        profit = max(profit, prices[i]-num);
           
           num = min(num,prices[i]);
        }
        
        return profit;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] arr) {
        int[] maxarr =  new int[arr.length];
        int[] minarr =  new int[arr.length];
        minarr[0]=arr[0];
        maxarr[arr.length-1]=arr[arr.length-1];
        for(int i=1;i<arr.length;i++){
            minarr[i]=Math.min(arr[i],minarr[i-1]);
        }
        for(int j=arr.length-2;j>-1;j--){
            maxarr[j]=Math.max(maxarr[j+1],arr[j]);
        }
        int ans=0;
        for(int i=0;i<arr.length;i++)ans=Math.max(ans,maxarr[i]-minarr[i]);
        return ans;
    }
}